### ISSUE TITLE
NIP-19 enhancement to include unsigned events

### ISSUE THREAD

**Issue by nQuiz** [nostr-protocol/nips#593]
## Problem

Private keys cannot be (safely) loaded into a browser, and there appears to be no easy-to-use, cross-device, cross-platform mechanism for signing events.

This is particularly problematic for developers who wish to integrate Nostr into **websites** accessed by **mobile users** (eg android)

The approach most widely used today (installation of chrome / firefox extensions) does not appear to be functional on mobile, and also requires a worrying "Read and Change all data on the websites you visit" permission.

Feedback is very welcome - perhaps this has already been discussed in a NIP / Issue I have overlooked?  The goal is mainly to enable Nostr users to sign events prepared by web browsers (eg Safari, Firefox, Brave) on both Android and iOS so that we can implement basic stuff like "sign in with nostr" and "reply to post" without ever having to ask for private keys.

## Reference Solution

This problem was solved for **lightning transactions** by invoking a URL such as `lightning:lnbc10n1pjgrtzfpp52jzf8tLNURLstring..`.

This is composed of a [scheme](https://www.rfc-editor.org/rfc/rfc2717) (`lightning:`) and a "scheme specific part" (the bit after the colon).

The browser uses the _scheme_ to launch the relevant _client application_ (eg Phoenix, WoS) in which the user confirms and broadcasts the payment.

## Proposed Solution

The Nostr solution could be much simpler than a lightning transaction (which requires a dedicated server).  

[NIP-21](/nostr-protocol/nips/blob/master/21.md) already defines a `nostr` scheme, and a set of prefixes in [NIP-19](/nostr-protocol/nips/blob/master/19.md)

This proposal is simply to add a new TLV type 0 / "special" bech32 prefix of `nsign`, containing the _unsigned event_.  As also discussed in [NIP-19](/nostr-protocol/nips/blob/master/19.md) this might contain Type 1 relays to which  to publish the signed event.

Our thin-client (web) application can then ask the fat-client (nostr) app - which typically does have the private key - to sign and publish the event.  This can be subsequently fetched from the relay(s) by the web app using a _tag_ as the identifier.

Mermaid diagram of the process flow:

```mermaid
flowchart TD
    A[Web App: request\nevent signature]
    A-->|1. Submit local bech32-encoded URL|B[Browser: Opens \n Client App]
    B-->|Passes bech32 data|C[Client App: decodes & signs event]
    C-->|Sends to Relay|D[Relay: Relays the event]
    A<-->|2. waits / polls|D
```

To illustrate with an event:

```
{
  "kind": 1,
  "tags": [
    [
      "e",
      "b3e392b11f5d4f28321cedd09303a748acfd0487aea5a7450b3481c60b6e4f87",
      "wss://relay.example.com"
    ]
  ],
  "content": "just sign here!"
}
```

The web app would convert the above, eg to: `nostr:nsign10elfcs4fr0l0r8af98jlmgdh9c8tcxjvz9qkw038js35mp4dma8qzvjptg`

The browser can then launch the users preferred `nostr` handler which - after user verification - can broadcast the _signed_ event (with `sig`, `id`, `created_at` and `pubkey`) to the requested / default relay(s).

The web app then fetches the signed event from the relay to confirm everything is as expected.

This article explains a bit more about adding handlers on android and apple: https://www.telerik.com/blogs/how-to-use-custom-url-schemes

Presuming the payload was not too large, the data might even be converted to a QR code, enabling air-gapped (and offline) signing of events.

**Comment by alexgleason** [comment]
Check out NIP-46 https://github.com/nostr-protocol/nips/blob/master/46.md

**Comment by nQuiz** [comment]
Thanks for this - do you know if there are any implementations (NIP-46 Signers) already?  Happy to close the issue if this is the way forward.

**Comment by alexgleason** [comment]
> do you know if there are any implementations (NIP-46 Signers) already?

Possibly this, I haven't tried it yet: https://github.com/nostr-dev-kit/ndk/tree/master/src/signers/nip46 

There's no native app I'm aware of yet. There needs to be.

**Comment by fiatjaf** [comment]
There is https://nsecbunker.com/

**Comment by nQuiz** [comment]
Thanks @fiatjaf - this project does look great, and we would definitely  like to use it ourselves (in two weeks obviously)

As I'm sure you appreciate, we're looking to cater towards users such as the 'normie grandma' whose son has helped her set up her profile in, say, Amethyst, and would now like to click the (big-font) 'sign in with Nostr' button on her new CleverPhone 

This reference implementation sounds great but there is no APK or recent activity: https://github.com/nostr-connect/nostrum

_Perhaps we will see NIP-46 support in the major Nostr Clients, such that they could be signers themselves?_

Obviously all this stuff is still new and we're excited to be part of this growth phase.  For now we'll just wait out and support users only on desktop (chrome / firefox with signing extension), as we don't want to support the use of private keys in browser apps for obvious reasons (even though the use of an extension still appears to be 'pasting a key into a browser' to most users)

**Comment by alexgleason** [comment]
I discovered a Nostr Connect native signer app does in fact exist. About to try it out: https://github.com/nostr-connect/nostrum

**Comment by pablof7z** [comment]
NsecBunker is available now, I just need to document it better and update its website, but search on Habla.news, there's a very comprehensive documentation on how to use it; I still need to document the CLI and the cli client implementation I wrote.

**Comment by nQuiz** [comment]
Thanks @pablof7z - I had followed your [guide](https://habla.news/a/naddr1qq9xuum9vd382mntv4eqzynhwden5te0wp6hyurvv4cxzeewv4eszrthwden5te0dehhxtnvdakqz8nhwden5te0dehhxarj94ex2mrp0yhxummtda6xzun09e3k7mgpzfmhxue69uhk7enxvd5xz6tw9ec82cspz3mhxue69uhhyetvv9ujuerpd46hxtnfduq3qamnwvaz7tmwdaehgu3wwa5kuegzypl4c26wfzswnlk2vwjxky7dhqjgnaqzqwvdvz3qwz5k3j4grrt46qcyqqq823cfpczge) and was able to run `docker run -d --name nsecbunkerd pablof7z/nsecbunkerd start --admin <your-npub>` however it wasn't clear how to extract the connection string.  Is there a command missing?  I did try messaging you on nostr last week (`npub1l2vyh47mk2p0qlsku7hg0vn29faehy9hy34ygaclpn66ukqp3afqutajft`)

In terms of this issue - one concern I have with NIP-46 is that it doesn't appear to be a 'one click' solution (or I've misunderstood something)

What we really need, on mobile browsers at least, is a simple button that will "do everything in one click".  Therefore it will surely need to open a (local / native) application to perform the signing (payload=unsigned event + relays) and the user can then use that app to decide which Nostr account to use to sign the event and broadcast to those relays.
